{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "14280d7f-fd9f-4de2-a11d-cb453601223a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pickle\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "sys.path.append('../src/')\n",
    "from configs import *\n",
    "from training import *\n",
    "from optimization import *\n",
    "from training import seed_everything, load_dataset\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "rcParams.update(fig_params)\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae25d332-a7f6-4c6a-acd8-27082de68ae7",
   "metadata": {},
   "source": [
    "## Load data and models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cca7c2f9-5476-4054-810e-69bd8106d2d9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "feature_list = ['physicochemical',\n",
    "                'AAC',\n",
    "                'CKSAAP type 1',\n",
    "                'TPC type 1',\n",
    "                'DPC type 1',\n",
    "                'DDE',\n",
    "                'GAAC',\n",
    "                'CKSAAGP type 1',\n",
    "                'GDPC type 1',\n",
    "                'GTPC type 1',\n",
    "                'Moran',\n",
    "                'Geary',\n",
    "                'NMBroto',\n",
    "                'CTDC',\n",
    "                'CTDT',\n",
    "                'CTDD',\n",
    "                'CTriad',\n",
    "                'KSCTriad',\n",
    "                'SOCNumber',\n",
    "                'QSOrder',\n",
    "                'PAAC',\n",
    "                'APAAC',\n",
    "                'ASDC',\n",
    "                'AC',\n",
    "                'CC',\n",
    "                'ACC',\n",
    "                'EAAC',\n",
    "                'EGAAC',\n",
    "                'AAIndex',\n",
    "                'BLOSUM62',\n",
    "                'ZScale'\n",
    "               ]\n",
    "\n",
    "comb, features = load_dataset(feature_list=feature_list, dataset_name='comb', features_folder=os.path.join('..', 'features'))\n",
    "## Train dataset\n",
    "train = comb[comb.Dataset=='train'].copy()\n",
    "test  = comb[comb.Dataset=='test'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9ff38b18-5fb5-4209-b125-debc1459a820",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Featurizer\n",
    "featurizer = Featurizer(data_folder=os.path.join('..', 'data'))\n",
    "\n",
    "## Model\n",
    "with open(os.path.join('..', 'models', 'LightCPP_20.pickle'), 'rb') as handle:\n",
    "    models = pickle.load(handle)\n",
    "best_iteration = models['best_iteration']\n",
    "n_models = 1\n",
    "\n",
    "## Selected features\n",
    "selected_features = ['CTDC_charge.G1',\n",
    "                     'CTDD_hydrophobicity_ARGP820101.1.residue50',\n",
    "                     'CKSAAP_IR.gap2',\n",
    "                     'CTDD_normwaalsvolume.2.residue100',\n",
    "                     'DDE_LA',\n",
    "                     'ASDC_AS',\n",
    "                     'PAAC_Xc1.V',\n",
    "                     'APAAC_Pc2.Hydrophobicity.1',\n",
    "                     'DDE_LP',\n",
    "                     'CKSAAP_LL.gap4',\n",
    "                     'APAAC_Pc1.A',\n",
    "                     'CTDD_hydrophobicity_FASG890101.2.residue50',\n",
    "                     'CKSAAGP_postivecharger.uncharger.gap1',\n",
    "                     'KSCTriad_g5.g5.g5.gap1',\n",
    "                     'ZScale_p1.z5',\n",
    "                     'QSOrder_Grantham.Xr.T',\n",
    "                     'ASDC_DM',\n",
    "                     'CKSAAP_TR.gap4',\n",
    "                     'PAAC_Xc1.G',\n",
    "                     'CKSAAGP_uncharger.aromatic.gap9']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f3f42af-8fe9-47d3-92f7-e256fa82b570",
   "metadata": {},
   "source": [
    "## Optimization loop"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6917e085-d24b-4b13-a2c5-452c05973b3f",
   "metadata": {},
   "source": [
    "WARNING: This loop takes hours to execute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "909fb476-1200-49ae-a5aa-0306520d395a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Negative_1 MILPTGPTSFKE\n",
      "Penetrability of target: 0.36327373596143514\n",
      "Anomaly of target: [1.09027296]\n",
      "Generation: 1 -- Best solution: MILPRNPPSFKD -- Best fitness: 0.33333 -- Prob_penetration: 0.89 -- Anomaly score: 1.01 -- Different letters from target: 4\n",
      "Generation: 2 -- Best solution: WILPTGPTSFKE -- Best fitness: 0.18108 -- Prob_penetration: 0.65 -- Anomaly score: 1.10 -- Different letters from target: 1\n",
      "Generation: 3 -- Best solution: MILPRGPTAFKE -- Best fitness: 0.14470 -- Prob_penetration: 0.80 -- Anomaly score: 1.01 -- Different letters from target: 2\n",
      "Generation: 4 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 5 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 6 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 7 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 8 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 9 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 10 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 11 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 12 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 13 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 14 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 15 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 16 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 17 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 18 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 19 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 20 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 21 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 22 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 23 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 24 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 25 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 26 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 27 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 28 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 29 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 30 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 31 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 32 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 33 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 34 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 35 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 36 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 37 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 38 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 39 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 40 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 41 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 42 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 43 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 44 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 45 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 46 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 47 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 48 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 49 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Generation: 50 -- Best solution: MILPRGPTSFKE -- Best fitness: 0.09524 -- Prob_penetration: 0.90 -- Anomaly score: 1.08 -- Different letters from target: 1\n",
      "Best solution with 0 different letters from target: MILPTGPTSFKE -- Prob_penetration: 0.36 -- Anomaly score: 1.09\n",
      "Best solution with 1 different letters from target: MILPRGPTSFKE -- Prob_penetration: 0.90 -- Anomaly score: 1.08\n",
      "Best solution with 2 different letters from target: LILPRGPTSFKE -- Prob_penetration: 0.90 -- Anomaly score: 1.08\n",
      "Best solution with 3 different letters from target: MIMPRGPTTFKE -- Prob_penetration: 0.89 -- Anomaly score: 1.06\n",
      "Best solution with 4 different letters from target: LILPRGPTAFKR -- Prob_penetration: 0.95 -- Anomaly score: 1.12\n",
      "Best solution with 5 different letters from target: LILPRNPPSFRE -- Prob_penetration: 0.82 -- Anomaly score: 1.04\n",
      "Best solution with 6 different letters from target: LILPRGPLTLEE -- Prob_penetration: 0.96 -- Anomaly score: 1.14\n",
      "Best solution with 7 different letters from target: YKLPRGPMAFRD -- Prob_penetration: 0.76 -- Anomaly score: 1.17\n",
      "Best solution with 8 different letters from target: RILPIMPLTSED -- Prob_penetration: 0.76 -- Anomaly score: 1.30\n",
      "Best solution with 9 different letters from target: MYRWRCPTMCLR -- Prob_penetration: 0.44 -- Anomaly score: 1.00\n"
     ]
    }
   ],
   "source": [
    "seed_everything(SEED)\n",
    "MAX_ITER = 50\n",
    "\n",
    "peptide_sequences = test[test['CPP'] == 0][['ID','Sequence']].copy()\n",
    "sequences = test[test['CPP'] == 0]['Sequence'].values\n",
    "results = pd.DataFrame(index=pd.MultiIndex.from_product([sequences, range(1, MAX_ITER+1)],\n",
    "                                                       names=['Input_peptide', 'Generation']),\n",
    "    columns=['Initial_Penetration','Best_solution','Best_fitness', 'Penetration_score', 'Anomaly_score', 'Different_letters'])\n",
    "\n",
    "for idx, row in peptide_sequences.iterrows():\n",
    "    ID, peptide_sequence = row.ID, row.Sequence\n",
    "    TARGET = peptide_sequence.strip()\n",
    "    print(ID, TARGET)\n",
    "    \n",
    "    # Set-up\n",
    "    len_gnome = len(TARGET)\n",
    "    max_iter = MAX_ITER\n",
    "    size_population = 500\n",
    "    genes = '''ARNDCEQGHILKMFPSTWYV'''\n",
    "\n",
    "    # Training features for anomaly detection\n",
    "    feat_train = train[selected_features]\n",
    "    clf_anomaly = LocalOutlierFactor(n_neighbors=20, novelty=True, contamination=0.1)\n",
    "    clf_anomaly.fit(feat_train)\n",
    "\n",
    "    # Sequence alignment settings\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "    aligner.target_internal_open_gap_score = -20\n",
    "    aligner.query_internal_open_gap_score = -20\n",
    "\n",
    "    generation = 1\n",
    "    population = []\n",
    "\n",
    "    optimization_info = {'Pop_size': size_population, 'max_iter': max_iter}\n",
    "\n",
    "    params_fit = {'obj': my_fitness, 'target_ligand': TARGET, 'aligner': aligner, 'model': models['refit_model'], 'clf_anomaly': clf_anomaly, \n",
    "                 'featurizer': featurizer, 'best_iteration': best_iteration, 'n_models': n_models}\n",
    "\n",
    "    features_TARGET = featurizer.compute_features(TARGET)\n",
    "    print(f\"Penetrability of target: {pred_penetration(features_TARGET, models['refit_model'], best_iteration, n_models)}\")\n",
    "    print(f\"Anomaly of target: {anomaly_score(features_TARGET, clf_anomaly)}\")\n",
    "\n",
    "    # Create initial population\n",
    "    max_diff_pct = 50  # set the percentage of different letters, user can change this value\n",
    "    for _ in range(size_population):\n",
    "        gnome = Individual.create_gnome(len_gnome, genes, TARGET, max_diff_pct)\n",
    "        population.append(Individual(gnome, params_fit))\n",
    "\n",
    "    # Define stopping criteria thresholds\n",
    "    penetration_threshold = 0.95\n",
    "    fitness_threshold = 0.1\n",
    "    anomaly_threshold = 1.3\n",
    "\n",
    "    best_sequences = {}\n",
    "\n",
    "    while generation <= max_iter:\n",
    "        # Sort the population in increasing order of fitness score\n",
    "        population = sorted(population, key=lambda x: x.fitness)\n",
    "\n",
    "        for individual in population:\n",
    "            if individual.num_diff_residues not in best_sequences:\n",
    "                best_sequences[individual.num_diff_residues] = individual\n",
    "            else:\n",
    "                if individual.fitness < best_sequences[individual.num_diff_residues].fitness:\n",
    "                    best_sequences[individual.num_diff_residues] = individual\n",
    "\n",
    "        # If the best solution in the population meets the stopping criteria, break the loop\n",
    "        best_solution = population[0]\n",
    "        features_best = featurizer.compute_features(''.join(best_solution.chromosome))\n",
    "        penetration_score = pred_penetration(features_best, models['refit_model'], best_iteration, n_models)\n",
    "        anomaly_score_value = anomaly_score(features_best, clf_anomaly)[0]\n",
    "\n",
    "        # Including the condition for num_diff_residues\n",
    "        if (penetration_score > penetration_threshold and anomaly_score_value < anomaly_threshold and best_solution.num_diff_residues <= 1):\n",
    "            break\n",
    "        # Including the condition for num_diff_residues\n",
    "        if (best_solution.num_diff_residues == 0):\n",
    "            break\n",
    "        # Generate new offsprings for new generation\n",
    "        new_generation = []\n",
    "\n",
    "        # Perform Elitism, that mean 10% of fittest population\n",
    "        # goes to the next generation\n",
    "        s = int((10*size_population) / 100)\n",
    "        new_generation.extend(population[:s])\n",
    "\n",
    "        # From 50% of fittest population, Individuals\n",
    "        # will mate to produce offspring\n",
    "        s = int((90*size_population) / 100)\n",
    "        for _ in range(s):\n",
    "            parent1 = random.choice(population[:50])\n",
    "            parent2 = random.choice(population[:50])\n",
    "            child = parent1.mate(parent2, genes)\n",
    "            new_generation.append(child)\n",
    "\n",
    "        population = new_generation\n",
    "\n",
    "        print(f\"Generation: {generation} -- \"\n",
    "              f\"Best solution: {''.join(population[0].chromosome)} -- \"\n",
    "              f\"Best fitness: {population[0].fitness:.5f} -- \"\n",
    "              f\"Prob_penetration: {penetration_score:.2f} -- \"\n",
    "              f\"Anomaly score: {anomaly_score_value:.2f} -- \"\n",
    "              f\"Different letters from target: {population[0].num_diff_residues}\"\n",
    "              )\n",
    "        \n",
    "        results.loc[(peptide_sequence, generation)]['Initial_Penetration'] = pred_penetration(features_TARGET, models['refit_model'], best_iteration, n_models)\n",
    "        results.loc[(peptide_sequence, generation)]['Best_solution'] = ''.join(population[0].chromosome)\n",
    "        results.loc[(peptide_sequence, generation)]['Best_fitness'] = population[0].fitness\n",
    "        results.loc[(peptide_sequence, generation)]['Penetration_score'] = penetration_score\n",
    "        results.loc[(peptide_sequence, generation)]['Anomaly_score'] = anomaly_score_value\n",
    "        results.loc[(peptide_sequence, generation)]['Different_letters'] = population[0].num_diff_residues\n",
    "        \n",
    "        generation += 1\n",
    "\n",
    "    # Print the best sequence for each number of different letters\n",
    "    for num_diff_residues, individual in sorted(best_sequences.items()):\n",
    "        print(f\"Best solution with {num_diff_residues} different letters from target: \"\n",
    "              f\"{''.join(individual.chromosome)} -- \"\n",
    "              f\"Prob_penetration: {pred_penetration(featurizer.compute_features(''.join(individual.chromosome)), models['refit_model'], best_iteration, n_models):.2f} -- \"\n",
    "              f\"Anomaly score: {anomaly_score(featurizer.compute_features(''.join(individual.chromosome)), clf_anomaly)[0]:.2f}\"\n",
    "              )\n",
    "    \n",
    "    results.to_pickle('../results/optimization_results.pickle')\n",
    "    \n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64526b0c-46f2-4d93-8f52-47c214eb65e2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
